### PR TITLE
update es query script

### PR DESCRIPTION
**Pull request by shaunahu**
This PR is to solve the ticket: https://github.com/aodn/backlog/issues/6062, with the objective of downloading all metadata records in Elasticsearch staging environment.

These changes have been made through this PR:
1. revise `search_es` function to download all 12k records
2. fix bugs in `get_predicted_keywords` code to use keyword in dict format
3. re-preprocess sample set, and re-train `development` model

By using raw data from staging environment, we now have 1856 records (after clean sample set), and 523 unique keywords. Tried different hypermeters, precision of 97% is the best result I could get (see the training log below). Will stop it here.
image

### CODE REVIEW

File: data_discovery_ai/pipeline.py [hunk 1]
```diff
@@ -115,11 +115,14 @@ def prepare_sampleSet(self, raw_data: pd.DataFrame) -> pd.DataFrame:
...
+        sampleSet_filtered = sampleSet[sampleSet["keywords"].apply(lambda x: x != [])]
+
         full_path = os.path.join(self.temp_dir, KEYWORD_SAMPLE_FILE)
 
-        preprocessor.save_to_file(sampleSet, full_path)
-        sampleSet = preprocessor.load_from_file(full_path)
-        return sampleSet
+        preprocessor.save_to_file(sampleSet_filtered, full_path)
+        sampleSet_filtered = preprocessor.load_from_file(full_path)
+        return sampleSet_filtered
```
Comment (vietnguyengit) [inline data_discovery_ai/pipeline.py:125]: I think better with the name: `filtered_sampleSet`

and why would you want to return the same thing after saving from loading again?

does this necessary? `sampleSet_filtered = preprocessor.load_from_file(full_path)`? as it adds up processing step

File: data_discovery_ai/utils/es_connector.py [hunk 2]
```diff
@@ -43,32 +43,72 @@ def connect_es(config: configparser.ConfigParser) -> Elasticsearch:
...
+    # get the first pit
+    first_pit_resp = client.open_point_in_time(
+        index=index,
+        keep_alive="1m",
+    )
+    first_pit = first_pit_resp["id"]
+
+    # the first search
+    first_query_body = {
+        "size": 100,
```
Comment (vietnguyengit) [inline data_discovery_ai/utils/es_connector.py:63]: hard-coded value `100`? why has to be `100`? and should it be declared in a global config/constant file?

File: data_discovery_ai/utils/es_connector.py [hunk 3]
```diff
@@ -43,32 +43,72 @@ def connect_es(config: configparser.ConfigParser) -> Elasticsearch:
...
+    df = pd.json_normalize(data)
+    dataframes.append(df)
+
+    # set current pit as the first one
+    current_pit = first_query_resp["pit_id"]
+
+    # conduct further search
+    for round in tqdm(range(1, rounds), desc="searching elasticsearch"):
+        qurry_body = {
+            "size": 100,
```
Comment (vietnguyengit) [inline data_discovery_ai/utils/es_connector.py:84]: same here, plus spelling mistake `qurry`

File: data_discovery_ai/utils/es_connector.py [hunk 4]
```diff
@@ -43,32 +43,72 @@ def connect_es(config: configparser.ConfigParser) -> Elasticsearch:
...
         df = pd.json_normalize(data)
         dataframes.append(df)
 
-        start_index += 1
+        # set current pit as the first one
+        current_pit = query_resp["pit_id"]
+
         round += 1
-        time.sleep(1)
+        time.sleep(10)
```
Comment (vietnguyengit) [inline data_discovery_ai/utils/es_connector.py:105]: why 10?

Comment (vietnguyengit) [reply data_discovery_ai/utils/es_connector.py:105]: and is hard-coded, maybe make it a parameter of `search_es()` method, together with the batch size `100` I commented above, should become

```python
def search_es(client: Elasticsearch, index: str = "es-indexer-staging", batch_size: int = 100, sleep_time: int = 10):
...
```

File: data_discovery_ai/utils/es_connector.py [hunk 5]
```diff
@@ -43,32 +43,72 @@ def connect_es(config: configparser.ConfigParser) -> Elasticsearch:
 
 def search_es(client: Elasticsearch):
     index = "es-indexer-staging"
```
Comment (vietnguyengit) [inline data_discovery_ai/utils/es_connector.py:45]: hard-coded index name here, should be a parameter.

File: data_discovery_ai/utils/es_connector.py [hunk 6]
```diff
@@ -43,32 +43,72 @@ def connect_es(config: configparser.ConfigParser) -> Elasticsearch:
...
+            "sort": [{"id.keyword": "asc"}],
+            "search_after": current_last_result,
+            "track_total_hits": False,
         }
+        query_resp = client.search(body=qurry_body)
 
-        resp = client.search(index=index, body=search_query)
+        data = query_resp["hits"]["hits"]
+        # set search after value
+        current_last_result = data[-1]["sort"]
```
Comment (vietnguyengit) [inline data_discovery_ai/utils/es_connector.py:95]: you might want to put this in a `try-except` block for exception handling, also use the `logger` to print out the exception error. Try to enable logging wherever important enough so we can monitor stuff on production in later stage, plus debugging needs.

File: data_discovery_ai/utils/es_connector.py [hunk 7]
```diff
@@ -43,32 +43,72 @@ def connect_es(config: configparser.ConfigParser) -> Elasticsearch:
...
 
-    raw_data = pd.concat(dataframes, ignore_index=True)
+    # close pit
+    resp = client.close_point_in_time(
+        id=current_pit,
+    )
+    print(resp)
 
+    raw_data = pd.concat(dataframes, ignore_index=True)
     return raw_data
```
Comment (vietnguyengit) [inline data_discovery_ai/utils/es_connector.py:114]: can be single line, `return pd.concat(dataframes, ignore_index=True)`

**Review by vietnguyengit — CHANGES_REQUESTED** [review]
please fix these comments first and I'll review again, thank @shaunahu

**Comment by shaunahu** [comment]
@vietnguyengit Thank you for your comments. They are solved by the latest commit with these changes:

1. Modified the hard-coded values to function parameters. An explanation of the default value setup has been added to the method comments. Personally, I don’t think adding these parameters to a global settings file is a good idea, as they are ad hoc and specific to this function. Including them in a global parameter file might confuse users who only reference the global settings.
2. Add try-except to some critical functions. I think there are some more functions I need to add, will do this along the unit test development.
3. Correct variable name

File: data_discovery_ai/utils/es_connector.py [hunk 8]
```diff
@@ -26,49 +26,98 @@
...
+    sleep_time: int = 5,
+):
     dataframes = []
-    batch_size = 100
-    resp = client.search(index=index, body={"query": {"match_all": {}}})
-    total_number = resp["hits"]["total"]["value"]
 
+    # get document count
+    count_resp = client.count(index=index)
+    total_number = count_resp["count"]
```
Comment (vietnguyengit) [inline data_discovery_ai/utils/es_connector.py:59]: what will happen if `count_resp` is None or `count_resp["count"]` is None?

File: data_discovery_ai/utils/es_connector.py [hunk 9]
```diff
@@ -26,49 +26,98 @@
...
     rounds = (total_number + batch_size - 1) // batch_size
```
Comment (vietnguyengit) [inline data_discovery_ai/utils/es_connector.py:60]: if there is error getting `total_number` as a valid number, `rounds` will fail.

File: data_discovery_ai/utils/es_connector.py [hunk 10]
```diff
@@ -26,49 +26,98 @@
...
 
-    # TODO: refactor the search query method.
-    for round in tqdm(range(rounds), desc="searching elasticsearch"):
-        search_query = {
+    # get the first pit
+    first_pit_resp = client.open_point_in_time(
+        index=index,
+        keep_alive="1m",
+    )
+    first_pit = first_pit_resp["id"]
```
Comment (vietnguyengit) [inline data_discovery_ai/utils/es_connector.py:67]: `first_pit_resp["id"]` always return valid value?

Comment (vietnguyengit) [reply data_discovery_ai/utils/es_connector.py:67]: and `first_pit_resp` as well

File: data_discovery_ai/utils/es_connector.py [hunk 11]
```diff
@@ -26,49 +26,98 @@
...
+        "size": batch_size,
+        "query": {"match_all": {}},
+        "pit": {"id": first_pit, "keep_alive": "1m"},
+        "sort": [{"id.keyword": "asc"}],
+    }
+    first_query_resp = client.search(body=first_query_body)
+
+    data = first_query_resp["hits"]["hits"]
+    # set search after value
+    current_last_result = data[-1]["sort"]
```
Comment (vietnguyengit) [inline data_discovery_ai/utils/es_connector.py:80]: same here, might need probably try-except or validation steps to ensure you get the right inputs for further steps.

File: data_discovery_ai/utils/es_connector.py [hunk 12]
```diff
@@ -26,49 +26,98 @@
...
+            "sort": [{"id.keyword": "asc"}],
+            "search_after": current_last_result,
+            "track_total_hits": False,
         }
+        query_resp = client.search(body=query_body)
+
+        try:
+            data = query_resp["hits"]["hits"]
+            # set search after value
+            current_last_result = data[-1]["sort"]
```
Comment (vietnguyengit) [inline data_discovery_ai/utils/es_connector.py:104]: needs validation too?

File: data_discovery_ai/utils/preprocessor.py [hunk 13]
```diff
@@ -251,15 +264,15 @@ def keywords_formatter(text: Union[str, List[dict]], vocabs: List[str]) -> List[
...
-            if keyword["title"] in vocabs and concept["id"] != "":
-                con = Concept(
-                    id=concept["id"].lower(),
-                    url=concept["url"],
-                    vocab_type=keyword["title"],
-                )
-                concept_str = con.to_json()
-                k_list.append(concept_str)
+        if keyword.get("concepts") is not None:
+            for con in keyword.get("concepts"):
```
Comment (vietnguyengit) [inline data_discovery_ai/utils/preprocessor.py:268]: use `concept` , not `con` for readability

**Review by vietnguyengit — CHANGES_REQUESTED** [review]
please review my comments

**Comment by vietnguyengit** [comment]
also @shaunahu , since this will be deployed in different cloud environments, some of your method's parameter values can be varied , e.g edge prod deployment downloads es data from prod es index, the cloud deployment LIKELY needs to reference the values from environment variables. so you might want to consider having your code able to read the values from environment variables. eg.

in `.env` file

```bash
ES_INDEX_NAME=blah-blah
```

in you code, where you pull data from ES, read the env variable for the index's name, also provide default value where the environment variable is missing, apply validation and throw exception and log the error.

**Comment by vietnguyengit** [comment]
> also @shaunahu , since this will be deployed in different cloud environments, some of your method's parameter values can be varied , e.g edge prod deployment downloads es data from prod es index, the cloud deployment LIKELY needs to reference the values from environment variables. so you might want to consider having your code able to read the values from environment variables. eg.
> 
> in `.env` file
> 
> ```shell
> ES_INDEX_NAME=blah-blah
> ```
> 
> in you code, where you pull data from ES, read the env variable for the index's name, also provide default value where the environment variable is missing, apply validation and throw exception and log the error.

not limited to this PR's scope, but also anywhere else that will use different parameter's values in different cloud deployment environments.

**Comment by vietnguyengit** [comment]
https://docs.python.org/3/library/os.html#os.getenv

File: data_discovery_ai/model/keywordModel.py [hunk 14]
```diff
@@ -294,7 +294,6 @@ def load_saved_model(trained_model: str) -> Optional[load_model]:
         saved_model = load_model(model_file_path, compile=False)
         return saved_model
     except Exception as e:
-        print(e)
         logger.info(
```
Comment (vietnguyengit) [inline data_discovery_ai/model/keywordModel.py:297]: should be error logging instead of info

File: data_discovery_ai/pipeline.py [hunk 15]
```diff
@@ -115,11 +115,13 @@ def prepare_sampleSet(self, raw_data: pd.DataFrame) -> pd.DataFrame:
...
 
+        # drop empty keywords rows
+        filtered_sampleSet = sampleSet[sampleSet["keywords"].apply(lambda x: x != [])]
+
         full_path = os.path.join(self.temp_dir, KEYWORD_SAMPLE_FILE)
 
-        preprocessor.save_to_file(sampleSet, full_path)
-        sampleSet = preprocessor.load_from_file(full_path)
-        return sampleSet
+        preprocessor.save_to_file(filtered_sampleSet, full_path)
```
Comment (vietnguyengit) [inline data_discovery_ai/pipeline.py:123]: might have an info logging here to say successfully saved the file?

Comment (shaunahu) [reply data_discovery_ai/pipeline.py:130]: `save_to_file` has the info logging already.

File: data_discovery_ai/model/keywordModel.py (hunk 14 above)
Comment (shaunahu) [reply data_discovery_ai/model/keywordModel.py]: loggings in expect block have been changed to error type in the latest commit

File: data_discovery_ai/utils/preprocessor.py (hunk 13 above)
Comment (shaunahu) [reply data_discovery_ai/utils/preprocessor.py]: changed to `concept` and `conceptObj` accordingly

File: data_discovery_ai/utils/es_connector.py (hunk 8 above)
Comment (shaunahu) [reply data_discovery_ai/utils/es_connector.py:59]: try-except and if condition for validation have been added to capture the fields raised from es response

**Comment by shaunahu** [comment]
@vietnguyengit Thank you for your comments. The above issues have been changed accordingly. I'll create another PR for moving environment variables in `.env` file.

**Comment by vietnguyengit** [comment]
wonderful, it is now flexible in a controlled manner, thanks @shaunahu

**Review by vietnguyengit — APPROVED** [review]

**Pull request merged**